### DISCUSSION TITLE
Use keyboard shortcuts

### DISCUSSION THREAD

**Discussion by pcar75** [jamulussoftware/jamulus#2213] [Ideas]
Bonjour,
I just read thread #2210 and keyboard shortcuts makes great sense.  For my part,  I would really welcome the option not to wrestle with the mouse and precision of graphic windows to mute or adjust the volume of a channel. This would reduce the amount of time my hands are off my instrument.

I suggest using the keyboard numeric keys to select the channel and the using the up/down keys to lower or raise the volume.  For ease of use, that would entail showing an identifying number near a channel (possibly the same as used in the file name of recorded tracks or logs entries on the server).

Thanks for considering this.

**Comment by pljones** [comment]
> numeric keys to select the channel 

You want to type "114" to select channel 114?  (Servers support up to 150 channels).  That seems cumbersome and error prone.  How do you select channel 1, channel 14, etc...

What if you want to adjust pan position?  Why restrict to fader level?

I'd suggest a keyboard shortcut to select the mixer, then Tab to move between channels until Enter is pressed, and then Tab to move between the channel controls, allowing cursor control of pan and fader independently, plus space bar to toggle the checkboxes.  Once a channel is set, Escape can move back a level.

That scheme gives full navigability.

I'd accept some _fast_ way to move rapidly from one channel to another - e.g. channel 7 to channel 41 - would be useful...  Suggestions welcome.  Maybe "hold down Alt and type, release Alt when you've entered all the digits"?

**Reply by henkdegroot** [reply]
> How do you select channel 1, channel 14, etc...

Well what about using a similar technique as used in many other device (like tv). 
Press a 1 and wait (thinking 1 sec wait)....now it selects channel 1.
Press a 1 and another number like 2 (within a second) and wait...now it selects channel 12
Press a 1 and a 2 and a 3, now select channel 123 (no more wait as we know 3 digits is the max).

Would be nice to have a solution which works using single keys (so you don't have to combine multiple keys or even need two hands).

In addition, the TAB to move from one channel to the next would be good to have as well.

**Comment by mslinn** [comment]
My 10keyless keyboards do not have a numeric keypad

**Comment by pcar75** [comment]
@pljones  

> You want to type "114" to select channel 114? (Servers support up to 150 channels). That seems cumbersome and error prone. How do you select channel 1, channel 14, etc...


I'm puzzled ?
Do you seriously consider presing the tab key 114 times or contineously is less error prone than typing 114 ?  The advantage I see for this would be to avoid displaying the channel number.

For my part, most of the servers I've played on  average a maximum of 20 users, which is seldom aurally satisfying. Possibly choirs do host more, but I doubt they allow 150 simultaneously, unless really exceptionnal network and computing power are at work; anyway I doubt one can view 150 simultaneously in a client.

To reiterate , my suggestion and priority is about minimizing my hands away from the instrument (guitar) while playing. My need is too adjust, ideally with one hand and rapidly,  a channel's overly loud volume, peak or noise,  or to mute the occasionnal (but rare) careless one(s).

Anything else can be set while I'm not playing (ex. panning) or at session setup, hence have time to deal with the mouse/touchpad and window precision.

**Reply by pljones** [reply]
> Do you seriously consider presing the tab key 114 times or contineously is less error prone than typing 114 ? The advantage I see for this would be to avoid displaying the channel number.

Did you read my suggestion?

But yes, the largest problem with navigating a large number of channels is _knowing what they're called_, regardless of how you address them - if you want to get there quickly rather than stepping through one at a time.  We don't insist any details of a player connection are unique (IP address:port number at the client end is 1-to-1 with channel number at the server end but neither scheme is really pleasant...).

**Comment by pcar75** [comment]
@henkdegroot    Having to wait for TV-like keys is one of thing I would avoid, as I'm playing and want my hands on the instrument.

**Reply by henkdegroot** [reply]
> my hands on the instrument.

So you won't be able to use the keyboard shortcuts at all :) Unless you link them to some kind of pageturner/footpedal.

**Reply by pcar75** [reply]
;-)  I did mean   hand(s) as less as possible off the instrument.  Though, I'd rather not add hardware but use the existing one ,
I find the idea and previous ones* interesting , although feet are also rather busy for drummers, piano/keyboard, etc.  :-/

*Thanks you for Sourceforge "Keyboard shortcuts in Jamulus"
https://sourceforge.net/p/llcon/discussion/533517/thread/f3252db0f5/

I also found interesting this 'Discussion' thread : Hardware Mixer/Mute/Solo Controls #2106.

**Comment by pcar75** [comment]
@mslinn    I would use the standard keyboard numeric keys , not a keypad as many (ex: laptop) do not have keypads.

**Comment by pcar75** [comment]
I have experience with Windows, a little with Linux but none with Mac or tablets and the like.  So, granted the OS has a way to switch focus to a desired application (ex.: Jamulus client), the rest of the key shortcut feature would apply to the client's 'console' window.

I suppose the client could be called with a switch (ex.: -s or -k) to enable channel number displaying and keyboard shortcut features.  When in focus and 'listening' for interaction, the 'console' window would interpret any number key as part of a channel number, any up/down keys as a notch up/down the volume level; possibly the letter 'm' or 'M' as mute.

Eventually, a nice keyboard shortcut sub-system could be developed in the future, even with default and personalized shortcuts as in most DAW.  If there's any development, I hope it would prioritize the musician's (incl. singers) focus on the music, less on the keyboard/mouse/etc..

Thanks.

**Comment by mslinn** [comment]
[Bome MIDI Translator Pro](https://www.bome.com/products/miditranslator) to the rescue!

**Reply by pcar75** [reply]
Thanks for the info,  I see no Linux version and not sure how it would use control inside an application.  I think there are some features in Windows OS to create helpful keyboard shortcuts / Hot keys and will look into that.  

Does anyone know if there are such features in Linux Ubuntu 20.4.x LTS ?
Thanks in advance!

**Reply by rdica** [reply]
```
Package: midisnoop
Architecture: amd64
Version: 0.1.2+git20141108.bc30f600187e-1build1
Priority: optional
Section: universe/sound
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Debian Multimedia Maintainers <debian-multimedia@lists.debian.org>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 642
Depends: libc6 (>= 2.14), libgcc-s1 (>= 3.0), libqt5core5a (>= 5.12.2), libqt5gui5 (>= 5.7.0) | libqt5gui5-gles (>= 5.7.0), libqt5widgets5 (>= 5.3.0), librtmidi4, libstdc++6 (>= 5)
Filename: pool/universe/m/midisnoop/midisnoop_0.1.2+git20141108.bc30f600187e-1build1_amd64.deb
Size: 170844
MD5sum: 1c450d676df29b84341a33a34999a91c
SHA1: 4f5f7e1d658b24da4440f86da03150b948219e09
SHA256: 996814721a181ca9353cba8aba9b174ec92ef89137a091857723a1050e530007
Homepage: https://github.com/surfacepatterns/midisnoop
Description-en: MIDI monitor and prober
 Can be used to monitor a MIDI device and/or software,
 and to send MIDI messages to a MIDI port
 to see how the device/software responds.
 .
 Support both ALSA and JACK MIDI ports via the RtMidi library.
Description-md5: ab408c0f86b7c8e403beb7268f16e32c
Task: ubuntustudio-audio
```
I use Ubuntu 18.04 and it has a similar program called `gmidimonitor`.

What these programs do is monitor and display midi messages over jack. Like any jack-aware program, you just connect the midi device ports to the monitor's ports and view the message traffic in the application. I doubt these programs have the bells and whistles that the commercial application listed above has, but it may be useful. I used it to give the info so I could script a cli volume controller for jack_mixer.

**Reply by pcar75** [reply]
@rdica   Thanks !  I hadn't thought of using MIDI but it might be an interesting avenue, like re-purposing a foot page turner suggested somewhere ... but it still might need a Jamulus client channel identifier (i.e. number) so it's selection and level , possibly other parameters ,  can be rapidly adjusted with one hand  :-|

**Reply by rdica** [reply]
A midi foot controller could be useful in a scenario you describe... :)

**Comment by pcar75** [comment]
Windows 10 Keyboard Shortcuts : 
No luck ; keyboard shortcuts / hot keys can select/start an application, possibly stop it, but cannot manipulate objects / functions inside the program. The PowerShell may have more facilities but unless fast channel identification/selection is available, an home-made utility probably won't allow me to be faster than when using a pointing device ...

**Comment by pcar75** [comment]
A depiction of the feature's possible behavior  ...
Edited
![image](https://user-images.githubusercontent.com/64435162/148658957-76faf737-eaaf-4218-aaff-cf4f286f6af6.png)

**Reply by pljones** [reply]
OK, so you type "11M" to mute channel 11?  And "113<up><up>" to increase the fader on channel 113?

If you wait over the timeout at any stage (or hit escape -- or ANY other key than one of those for the channel controls - including already mapped shortcuts), you have to start typing a number again.

Seems logically sound now.

It's another bit of overhead for the client to run as a separate task on top of everything else, though -- we don't manage the UI, that's done entirely by Qt currently.  So this would have to sit "in front of" the Qt layer.

So whether it's technically achievable with the current architecture, I don't know.

Although, saying that, I think there is a built-in Qt key monitoring system.

**Reply by pcar75** [reply]
> And "113" to increase the fader on channel 113?

And the up arrow , yes. 

The idea is to use for a subset of controls, the easiest way possible to manipulate them , a limited number of one finger strokes of one hand in this case.

Some caveat though,  is that rapidity of execution is still relative , depending on the server and the links latency as it is the server that will make the adjustments.  So, for example, it would perhaps be inappropriate to 'gauge' the levels (volume, panning, reverb, etc.) by notches finer than a tenth of max value.  If ones has to wait the return of the audio from the server to verify and readjust the level more precisely it kind of defeats the purpose of quick adjustment I seek.

It take it that Qt is a framework or API; it would probably recognize more than pointing device.

I realize that this is not an easy modification as it entails the client and sever aspect and communication between.

EDIT : The modification may not be as big as I initially thought as Alt+M is already implemented to 'Mute Myself'.

**Comment by pcar75** [comment]
As an eventual ( if / when ? ) implementation of my suggestion ... one should consider that the identification of channels is already available when the `--ctrlmidich` option is used.  I have used this option with minimal but syntax-valid parameters (i.e. `--ctrlmidich "1;o2"` ) to obtain channel-number-prefixed names with faders even though I use no midi device.  

![Jamulus Client_display with Channel numbers_02](https://user-images.githubusercontent.com/64435162/176751184-f6b07d6e-807a-49b5-b87c-b7c85f16ef0f.png)

But beware that this option is currently causing a bug in the jamulus.ini file.
Ref. Issues / "Option ctrlmidich writes fader tags in .ini file as Channel':'Name #2680"
EDIT: As of 2022-07-05, this bug's resolution is slated for v3.9.1.

**Comment by JaminShanti** [comment]
+1 for Mute button.  How many times I have disconnected myself hitting Spacebar.
